stress/src/metrics_cached_attrs.rs

```diff
@@ -0,0 +1,53 @@
+use lazy_static::lazy_static;
+use opentelemetry::AttributeSet;
+use opentelemetry::{
+    metrics::{Counter, MeterProvider as _},
+    KeyValue,
+};
+use opentelemetry_sdk::metrics::{ManualReader, SdkMeterProvider};
+use rand::{rngs::SmallRng, Rng, SeedableRng};
+use std::borrow::Cow;
+
+mod throughput;
+
+lazy_static! {
+    static ref PROVIDER: SdkMeterProvider = SdkMeterProvider::builder()
+        .with_reader(ManualReader::builder().build())
+        .build();
+    static ref ATTRIBUTE_VALUES: [&'static str; 10] = [
+        "value1", "value2", "value3", "value4", "value5", "value6", "value7", "value8", "value9",
+        "value10"
+    ];
+    static ref COUNTER: Counter<u64> = PROVIDER
+        .meter(<&str as Into<Cow<'static, str>>>::into("test"))
+        .u64_counter("hello")
+        .init();
+    static ref ATTRIBUTE_SETS: Vec<AttributeSet> = {
+        let mut vec = Vec::new();
+        for i0 in 0..ATTRIBUTE_VALUES.len() {
+            for i1 in 0..ATTRIBUTE_VALUES.len() {
+                for i2 in 0..ATTRIBUTE_VALUES.len() {
+                    vec.push(AttributeSet::from(&[
+                        KeyValue::new("attribute1", ATTRIBUTE_VALUES[i0]),
+                        KeyValue::new("attribute2", ATTRIBUTE_VALUES[i1]),
+                        KeyValue::new("attribute3", ATTRIBUTE_VALUES[i2]),
+                    ]))
+                }
+            }
+        }
+
+        vec
+    };
+}
+
+fn main() {
+    throughput::test_throughput(test_counter);
+}
+
+fn test_counter() {
+    let mut rng = SmallRng::from_entropy();
+    let len = ATTRIBUTE_SETS.len();
+    let index = rng.gen_range(0..len);
+
+    COUNTER.add(1, ATTRIBUTE_SETS[index].clone());
```

Comment:
Can you please confirm if the change in this PR is backward incompatible (breaking) change for existing instrumentation (as mentioned in the PR description)? As there are no changes done in the stress/metrics.rs, and it seems to be building correctly - https://github.com/open-telemetry/opentelemetry-rust/blob/27d338dd954ef7681242ead47a8a61c964dcc428/stress/src/metrics.rs#L38-L45

Reply:
With Julian's `ToKv` trait suggestion this works in a backwards compatible way, since a `&[KeyValue]` implements the `ToKv` trait, it automatically gets coerced into an interator of `ToKv` values.

Reply:
Thanks for clarifying. So, the below code will still work as before, but there won't be any perf improvement in this case as a new instance of `(Internal)AttributeSet` gets created in each call to counter::add(). 

```rust
    // Create a vector of KeyValue pairs.
    let key_values = vec![
        KeyValue::new("mykey1", "myvalue1"),
        KeyValue::new("mykey2", "myvalue2"),
    ];

    // Record measurements using the Counter instrument.
    counter.add(10, key_values.as_ref());
    counter.add(10, key_values.as_ref());
```

While supported, this is not the recommended approach and is good to be documented (probably as separate PR).

Reply:
Actually, that specific example is backward compatible, but observables **are not** backward compatible.  Those must take an `AttributeSet` in because they rely on boxed traits instead of generics, and thus cannot be easily coerced without a bunch of overhead.